internal/handlers/pg/pgdb/indexes.go

```diff
@@ -22,25 +22,61 @@ import (
 	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 )
 
-// indexParams describes the parameters for creating an index.
+// indexParams contains parameters for creating an index.
+// TODO This type will become exported in https://github.com/FerretDB/FerretDB/issues/1509 (similar to QueryParams).
 type indexParams struct {
-	schema   string // pg schema name
-	table    string // pg table name
-	isUnique bool   // whether the index is unique
+	db         string   // FerretDB database name
+	collection string   // FerretDB collection name
+	index      string   // FerretDB index name
+	key        indexKey // Index specification (pairs of field names and sort orders) // TODO
+	unique     bool     // Whether the index is unique
+}
+
+// indexKey defines a type for index key - pairs of field names and sort orders.
+type indexKey []indexKeyPair
+
+// indexKeyPair consists of a field name and a sort order that are part of the index.
+type indexKeyPair struct {
+	field string
+	order indexOrder
+}
+
+// indexOrder defines a type for index sort order.
+type indexOrder int8
+
+// indexOrder constants.
+const (
+	indexOrderAsc  indexOrder = 1
+	indexOrderDesc indexOrder = -1
+)
+
+// createIndex creates a new index for the given params.
+// TODO This method will become exported in https://github.com/FerretDB/FerretDB/issues/1509.
+func createIndex(ctx context.Context, tx pgx.Tx, ip *indexParams) error {
+	pgTable, pgIndex, err := newMetadata(tx, ip.db, ip.collection).setIndex(ctx, ip.index, ip.key, ip.unique)
+	if err != nil {
+		return err
+	}
+
+	if err := createIndexIfNotExists(ctx, tx, ip.db, pgTable, pgIndex, true); err != nil {
+		return err
+	}
+
+	return nil
 }
 
 // createIndexIfNotExists creates a new index for the given params if it does not exist.
-func createIndexIfNotExists(ctx context.Context, tx pgx.Tx, p *indexParams) error {
+func createIndexIfNotExists(ctx context.Context, tx pgx.Tx, schema, table, index string, isUnique bool) error {
```

Comment:
It seems that the list of arguments here is pretty long. I see that we modified the `indexParams` to describe the FerretDB layer logic, and not the PostgreSQL for the `createIndex` function. If I understand, we don't plan to use this function in lot of places and it will be rather something internal for `pgdb`, so I guess it's not a big problem.

However the naming here looks a bit misleading to me. There's `createIndexIfNotExists` function, which creates an index based on PostgreSQL related arguments.

Afterwards we have `createIndex`, which is a wrapper on top of `createIndexIfNotExists`, it will be the exported function and (If I understand) preferably used in our code.

Both functions create index if it doesn't exist, but they differ from each other only by "If not exists" part.

I would suggest to differentiate those function based on their layers, or - just keep the name the same, but make one exported (`createIndex` for pgdb layer and `CreateIndex`for FerretDB layer).

What do you think?

Reply:
I renamed the unexported one to `createPGIndexIfNotExists` to clarify that it works with PG. The future-exported `CreateIndex` is needed to work with index as FerretDB abstraction (metadata + PG index).

I also renamed `createTableIfNotExists` to `createPGTableIfNotExists` to keep the naming consistent.

Reply:
Looks good!